plugins/out_s3/s3.c

```diff
@@ -1098,6 +1098,7 @@ static int s3_put_object(struct flb_s3 *ctx, const char *tag, time_t create_time
             return -1;
         }
         final_body = (char *) compressed_body;
+        flb_free(compressed_body);
```

Comment:
this doesn't work. The line above does not copy memory, it copies the address.

You need to write code after the call to S3 API which frees the compressed body IFF compression is turned on. Some solution like that.